src/compiler/checker.ts

```diff
@@ -20367,6 +20367,28 @@ namespace ts {
                 // Compare the remaining non-discriminant properties of each match.
                 let result = Ternary.True;
                 for (const type of matchingTypes) {
+                    if (getObjectFlags(source) & ObjectFlags.Reference && getObjectFlags(type) & ObjectFlags.Reference && (source as TypeReference).target === (type as TypeReference).target &&
+                        !isTupleType(source) && !(isMarkerType(source) || isMarkerType(type))) {
+                        // When strictNullChecks is disabled, the element type of the empty array literal is undefinedWideningType,
+                        // and an empty array literal wouldn't be assignable to a `never[]` without this check.
+                        if (isEmptyArrayLiteralType(source)) {
+                            return Ternary.True;
+                        }
+                        // We have type references to the same generic type, and the type references are not marker
+                        // type references (which are intended by be compared structurally). Obtain the variance
+                        // information for the type parameters and relate the type arguments accordingly.
+                        const variances = getVariances((source as TypeReference).target);
+                        // We return Ternary.Maybe for a recursive invocation of getVariances (signalled by emptyArray). This
+                        // effectively means we measure variance only from type parameter occurrences that aren't nested in
+                        // recursive instantiations of the generic type.
+                        if (variances === emptyArray) {
+                            return Ternary.Unknown;
+                        }
+                        if (!typeArgumentsRelatedTo(getTypeArguments(source as TypeReference), getTypeArguments(type as TypeReference), variances, /* reportErrors */ false, IntersectionState.None)) {
+                            return Ternary.False;
+                        }
+                    }

```

Comment:
This is effectively a copy-paste (with very small adjustments) of the block here:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L20150-L20171

This definitely should be deduplicated if the solution itself is correct. I will wait for the review before making any changes to this.

Reply:
Using variance results more is usually good for perf, but a variance-based result and a structural result being different is fundamentally the problem. Usually means some comparison needs to add an `Unmeasurrable` or `Unreliable` marker to the result calculated for the variance relationship.

Reply:
Do you think there is a problem with the added code? Or that there is a problem elsewhere and that elsewhere Unmeasurable/Unreliable is not handled correctly?

Reply:
More likely that a non-strictly-co--or-contra-variant comparison somewhere needs to throw out a `Unmeasurable` or `Unreliable` flag but isn't yet doing so.